File: src/main/resources/harness/workspace/imports/Greeter.java
package com.example.util;

import java.util.ArrayList;
import java.util.List;

/**
 * Middle of the chain — uses {@link Formatter} with <b>no import at all</b>.
 *
 * <p>That is the point of this one. Same-package resolution does not go through an import statement, so
 * it exercises a different path in the name environment: ECJ asks for {@code com.example.util.Formatter}
 * directly rather than resolving an import first. A build where imports worked and same-package lookup
 * did not would look completely fine in {@code Main} and fail only here.</p>
 *
 * <p><b>Things worth trying.</b> Type {@code Formatter.} anywhere below and the completion popup should
 * offer {@code shout} and {@code quiet} — members of another file, with their real signatures. Ctrl+B on
 * {@code Formatter} should open it. Hover should quote its javadoc.</p>
 */
public class Greeter {

    private final String greeting;

    public Greeter(String greeting) {
        this.greeting = greeting == null ? "hello" : greeting;
    }

    /** The one everybody calls. Reaches {@code Formatter} without naming its package. */
    public String greet(String who) {
        return Formatter.shout(greeting + ", " + who);
    }

    /** A generic signature, so the type shown in {@code Main} is worth reading. */
    public List<String> greetAll(List<String> names) {
        List<String> out = new ArrayList<>();
        for (String name : names) {
            out.add(greet(name));
        }
        return out;
    }

    /** A static factory, so {@code Main} can reach this without a constructor. */
    public static Greeter casual() {
        return new Greeter("hey");
    }

    /** Deliberately quiet — {@code Formatter.quiet} is only reachable from here. */
    public String murmur(String who) {
        return Formatter.quiet(greeting + ", " + who);
    }
}

File: src/main/resources/harness/workspace/imports/Greeter.js
// Middle of the chain — a module that itself imports a module.
//
// This is the file that decided the implementation. Rhino ships CommonJS and it would have been the
// natural thing to bind our `import` statement to, except that `Require` builds each module's scope
// internally with no hook to inject a binding into — so a module with an import of its own could not be
// served at all. Which is exactly this file, and it is the ordinary shape rather than an edge case.
//
// Things worth trying:
//   * Type `Formatter.` below — the popup should offer `shout` and `quiet`, read out of the other file.
//   * Ctrl+B on `Formatter` opens it.
//   * Delete the import line and `Formatter` should go unresolved — then undo.
//   * `module.exports = { ... }` is read too; try replacing the two `exports.x =` lines with one object
//     and check App.js still completes.

import util.Formatter;

exports.greet = function (who) {
    return Formatter.shout('hello, ' + who);
};

exports.murmur = function (who) {
    return Formatter.quiet('hello, ' + who);
};

// A plain value export, so the popup has something that is not a function in it.
exports.defaultName = 'world';
